File: kotlin/src/main/kotlin/advertisements/user/domain/MemberUser.kt
package advertisements.user.domain

import advertisements.shared.value_object.CivicCenterId
import advertisements.shared.value_object.Email
import advertisements.shared.value_object.Password
import advertisements.shared.value_object.UserId
import advertisements.user.domain.value_object.Role
import advertisements.user.domain.value_object.UserStatus
import advertisements.user.domain.value_object.MemberNumber
import advertisements.user.domain.exceptions.InvalidUserException

class MemberUser private constructor(
    id: UserId,
    email: Email,
    role: Role,
    private val memberNumber: MemberNumber,
    private val civicCenterId: CivicCenterId,
    status: UserStatus
) : UserBase(id, email, role, status) {

    init {
        if (role != Role.MEMBER) {
            throw InvalidUserException.build("Invalid role for member user")
        }
    }

    companion object {
        @Throws(InvalidUserException::class)
        fun signUp(
            id: UserId,
            email: Email,
            password: Password,
            role: Role,
            memberNumber: MemberNumber,
            civicCenterId: CivicCenterId
        ): MemberUser {
            val member = MemberUser(id, email, role, memberNumber, civicCenterId, UserStatus.ENABLED)
            member.password = password
            return member
        }

        @Throws(InvalidUserException::class)
        fun fromDatabase(
            id: UserId,
            email: Email,
            role: Role,
            memberNumber: MemberNumber,
            civicCenterId: CivicCenterId,
            status: UserStatus
        ): MemberUser {
            return MemberUser(id, email, role, memberNumber, civicCenterId, status)
        }
    }

    fun disable() {
        status = UserStatus.DISABLED
    }

    fun enable() {
        status = UserStatus.ENABLED
    }

    fun memberNumber(): MemberNumber {
        return memberNumber
    }

    fun civicCenterId(): CivicCenterId {
        return civicCenterId
    }
}

File: typescript/src/advertisements/user/domain/MemberUser.ts
import { UserBase } from "./UserBase";
import { UserId } from "../../shared/domain/value-object/UserId";
import { Email } from "../../shared/domain/value-object/Email";
import { Role } from "./value-object/Role";
import { UserStatus } from "./value-object/UserStatus";
import { InvalidUserException } from "./exceptions/InvalidUserException";
import { MemberNumber } from "./value-object/MemberNumber";
import { Password } from "../../shared/domain/value-object/Password";
import {CivicCenterId} from "../../shared/domain/value-object/CivicCenterId";

export class MemberUser extends UserBase {
    private _memberNumber: MemberNumber;
    private _civicCenterId: CivicCenterId;

    /** @throws InvalidUserException */
    protected constructor(id: UserId, email: Email, role: Role, memberNumber: MemberNumber, civicCenterId: CivicCenterId, status: UserStatus) {
        if (role !== Role.MEMBER) {
            throw InvalidUserException.build('Invalid role for member user');
        }

        super(id, email, role, status);
        this._memberNumber = memberNumber;
        this._civicCenterId = civicCenterId;
    }

    /** @throws InvalidUserException */
    public static signUp(id: UserId, email: Email, password: Password, role: Role, memberNumber: MemberNumber, civicCenterId: CivicCenterId): MemberUser {
        const member = new MemberUser(id, email, role, memberNumber, civicCenterId, UserStatus.ENABLED);
        member._password = password;
        return member;
    }

    public disable(): void {
        this._status = UserStatus.DISABLED;
    }

    public enable(): void {
        this._status = UserStatus.ENABLED;
    }

    /** @throws InvalidUserException */
    public static fromDatabase(id: UserId, email: Email, role: Role, memberNumber: MemberNumber, civicCenterId: CivicCenterId, status: UserStatus): MemberUser {
        return new MemberUser(id, email, role, memberNumber, civicCenterId, status);
    }

    public memberNumber(): MemberNumber {
        return this._memberNumber;
    }

    public civicCenterId(): CivicCenterId {
        return this._civicCenterId;
    }
}
